### DISCUSSION TITLE
How to specify an element is a text one?

### DISCUSSION THREAD

**Discussion by ooker777** [syntax-tree/unist#137] [Q&A] [answered]
This code:
```ts
import { fromHtml } from "hast-util-from-html";
import { select } from "hast-util-select";

const tree = fromHtml("<h1>Hi</h1>>", { fragment: true });
console.log(select("h1", tree)!.children[0].value);
```
creates this error
```
Property 'value' does not exist on type 'ElementContent'.
  Property 'value' does not exist on type 'Element'.
```
How can I fix that?

Also, why is the text considered as a child of the tag? Shouldn't the children property only contain other elements? The `value` property can move up to the object.

**Comment by wooorm** [comment, marked as answer]
TypeScript gives you a type error because you have a type error.
If you want types, use TS. TS will give you errors when you make problems. This is that.

How to use TS:

You store things in variables, and perform checks on them:

```js
const tree = fromHtml("<h1>Hi</h1>>", { fragment: true }); // Root
const h1 = select("h1", tree) // Element
if (!h1) throw new Error("main heading missing")
const head = h1.children[0] // ElementContent
if (head.type !== 'text') throw new Error("there are non-text things in the head!")
console.log(head.value)
```

However, do not do you want you do. Your goal is to get the text of a heading. Headings can contain other things than text (as the types indicate). If you want text, use a tool for that:

```js
import {toText} from 'hast-util-to-text'

const tree = fromHtml("<h1>Hi</h1>>", { fragment: true }); // Root
const h1 = select("h1", tree) // Element
if (!h1) throw new Error("main heading missing")
const text = toText(h1) // string
```

See also:

- https://unifiedjs.com/learn/recipe/narrow-node-typescript/
- https://unifiedjs.com/learn/guide/syntax-trees-typescript/
- https://unifiedjs.com/learn/recipe/tree-traversal-typescript/

> Shouldn't the children property only contain other elements?

No, that is not needed. There is no such restriction.

> The value property can move up to the object.

Anything could be done. But some things are not good things. This is not a good thing. There would be different sources of truth. Which will bite everyone in their butt.
